$OpenBSD: patch-frontends_php_acknow_php,v 1.1.2.1 2011/08/19 13:41:13 jasper Exp $

Security fix for CVE-2011-2904,
Zabbix "backurl" Cross-Site Scripting Vulnerability

Fix from upstream svn -r20789.

--- frontends/php/acknow.php.orig	Fri Apr 15 21:17:31 2011
+++ frontends/php/acknow.php	Thu Aug 18 20:03:38 2011
@@ -28,6 +28,7 @@ $page['title']	= 'S_ACKNOWLEDGES';
 $page['file']	= 'acknow.php';
 $page['hist_arg'] = array('eventid');
 
+ob_start();
 include_once('include/page_header.php');
 
 ?>
@@ -52,10 +53,11 @@ include_once('include/page_header.php');
 	);
 	check_fields($fields);
 
+	$_REQUEST['backurl'] = get_request('backurl', 'tr_status.php');
+
 	if(isset($_REQUEST['cancel'])){
-		$url = new CUrl(urldecode($_REQUEST['backurl']));
-		jsRedirect($url->getUrl());
-		exit();
+		ob_end_clean();
+		redirect($_REQUEST['backurl']);
 	}
 
 	if(!isset($_REQUEST['events']) && !isset($_REQUEST['eventid']) && !isset($_REQUEST['triggers'])){
@@ -64,7 +66,6 @@ include_once('include/page_header.php');
 	}
 
 	$bulk = !isset($_REQUEST['eventid']);
-	$_REQUEST['backurl'] = get_request('backurl', 'tr_status.php');
 ?>
 <?php
 	if(!$bulk){
@@ -113,12 +114,13 @@ include_once('include/page_header.php');
 		}
 
 		if(isset($_REQUEST['saveandreturn'])){
-			$url = new CUrl(urldecode($_REQUEST['backurl']));
-			jsRedirect($url->getUrl());
-			exit();
+			ob_end_clean();
+			redirect($_REQUEST['backurl']);
 		}
  	}
 
+ob_end_flush();
+
 ?>
 <?php
 	$msg = $bulk ? ' BULK ACKNOWLEDGE ' : expand_trigger_description_by_data($event_trigger);
@@ -187,13 +189,10 @@ include_once('include/page_header.php');
 	$frmMsg->addRow(S_MESSAGE, new CTextArea('message', '', 80, 6));
 	$frmMsg->addItemToBottomRow(new CButton('saveandreturn', $btn_txt2));
 	$bulk ? '' : $frmMsg->addItemToBottomRow(new CButton('save', $btn_txt));
-	$frmMsg->addItemToBottomRow(new CButtonCancel('&backurl='.$_REQUEST['backurl']));
+	$frmMsg->addItemToBottomRow(new CButtonCancel('&backurl='.urlencode($_REQUEST['backurl'])));
 
 	$frmMsg->show(false);
 
-?>
-<?php
 
 include_once('include/page_footer.php');
-
 ?>
